clang/lib/Sema/SemaInit.cpp

```diff
@@ -268,6 +268,69 @@ static void CheckStringInit(Expr *Str, QualType &DeclT, const ArrayType *AT,
   updateStringLiteralType(Str, DeclT);
 }
 
+template <class Predicate, class RecordFilter>
+std::enable_if_t<std::is_void_v<decltype(std::declval<Predicate &>()(
+                     std::declval<const FieldDecl *>()))>,
+                 size_t>
+forEachFieldRecursive(const RecordDecl *Record, Predicate FieldPred,
+                      RecordFilter RecordPreFilter) {
+  size_t NumFound = 0;
+  SmallVector<const RecordDecl *> Stack;
+  llvm::SmallDenseSet<const RecordDecl *> Seen;
+  Stack.push_back(Record);
+  while (!Stack.empty()) {
+    const RecordDecl *RD = Stack.back();
+    Stack.pop_back();
+    const size_t OldStackSize = Stack.size();
+    if (RD) {
+      RD = dyn_cast_if_present<RecordDecl>(RD->getCanonicalDecl());
```

Comment:
Curleys shouldn't be needed here.

IIRC, `getCanonicalDecl` shouldn't return null here if RD does, so perhaps the suggestion?
```suggestion
      RD = dyn_cast_if_present<RecordDecl>(RD ? RD : RD->getCanonicalDecl());
```

And do away with the 'if'? Since all we're doing is preserving canonicalization.